pydantic/main.py

```diff
@@ -363,11 +373,23 @@ def __setattr__(self, name, value):
         self.__fields_set__.add(name)
 
     def __getstate__(self) -> 'DictAny':
-        return {'__dict__': self.__dict__, '__fields_set__': self.__fields_set__}
+        return {
+            '__dict__': self.__dict__,
+            '__fields_set__': self.__fields_set__,
+            '__private_attributes_values__': {k: getattr(self, k, Undefined) for k in self.__private_attributes__},
+        }
 
     def __setstate__(self, state: 'DictAny') -> None:
-        object.__setattr__(self, '__dict__', state['__dict__'])
-        object.__setattr__(self, '__fields_set__', state['__fields_set__'])
+        object_setattr(self, '__dict__', state['__dict__'])
+        object_setattr(self, '__fields_set__', state['__fields_set__'])
+        self._set_private_attributes(state['__private_attributes_values__'], need_copy=False, check=False)
+
+    def _set_private_attributes(self, source: 'DictStrAny', need_copy: bool = True, check: bool = True) -> None:
+        for name, value in source.items():
+            if value is not Undefined:
+                object_setattr(self, name, smart_deepcopy(value) if need_copy else value)
+            elif check and not hasattr(self, name):
+                raise AttributeError(f'private attribute "{name}" is unset')
```

Comment:
maybe if check is `True` we should do the check first so the exception message is always the same?

Reply:
or make it identical here.

Reply:
Didn't get this one, could you explain?